streams/src/test/java/org/apache/kafka/streams/integration/RegexSourceIntegrationTest.java

```diff
@@ -0,0 +1,353 @@
+        produceMessage(TOPIC_C, Arrays.asList(topicCTestMessage), producerConfig);
+        produceMessage(TOPIC_Y, Arrays.asList(topicYTestMessage), producerConfig);
+        produceMessage(TOPIC_Z, Arrays.asList(topicZTestMessage), producerConfig);
+
+        Properties consumerConfig = getConsumerConfig();
+
+        List<String> expectedReceivedValues = Arrays.asList(topicATestMessage, topic1TestMessage, topic2TestMessage, topicCTestMessage, topicYTestMessage, topicZTestMessage);
+        List<KeyValue<String, String>> receivedKeyValues = IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig, DEFAULT_OUTPUT_TOPIC, 6);
+        List<String> actualValues = new ArrayList<>(6);
+
+        for (KeyValue<String, String> receivedKeyValue : receivedKeyValues) {
+            actualValues.add(receivedKeyValue.value);
+        }
+
+        streams.close();
+        Collections.sort(actualValues);
+        Collections.sort(expectedReceivedValues);
+        assertThat(actualValues, equalTo(expectedReceivedValues));
+    }
+
+    @Test(expected = AssertionError.class)
+    public void testNoMessagesSentExceptionFromOverlappingPatterns() throws Exception {
+
+        String fooMessage = "fooMessage";
+        String fMessage = "fMessage";
+
+
+        final Serde<String> stringSerde = Serdes.String();
+
+        Properties streamsConfiguration = getStreamsConfig();
+
+        KStreamBuilder builder = new KStreamBuilder();
+
+        /*
+          overlapping patterns here, no messages should be sent as TopologyBuilderException
+          will be thrown when the processor topology is built.
+         */
+        KStream<String, String> pattern1Stream = builder.stream(Pattern.compile("foo.*"));
+        KStream<String, String> pattern2Stream = builder.stream(Pattern.compile("f.*"));
+
+
+        pattern1Stream.to(stringSerde, stringSerde, DEFAULT_OUTPUT_TOPIC);
+        pattern2Stream.to(stringSerde, stringSerde, DEFAULT_OUTPUT_TOPIC);
+
+
+        // Remove any state from previous test runs
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfiguration);
+
+        KafkaStreams streams = new KafkaStreams(builder, streamsConfiguration);
+        streams.start();
+
+        Properties producerConfig = getProducerConfig();
+
+        produceMessage(FA_TOPIC, Arrays.asList(fMessage), producerConfig);
+        produceMessage(FOO_TOPIC, Arrays.asList(fooMessage), producerConfig);
+
+        Properties consumerConfig = getConsumerConfig();
+
+        try {
+            IntegrationTestUtils.waitUntilMinKeyValueRecordsReceived(consumerConfig, DEFAULT_OUTPUT_TOPIC, 2, 5000);
+        } finally {
+            streams.close();
+        }
+
+    }
+
+    private void produceMessage(String inputTopic, List<String> input, Properties producerConfig) throws Exception {
+        IntegrationTestUtils.produceValuesSynchronously(inputTopic, input, producerConfig);
+    }
+
+
+    private Properties getProducerConfig() {
```

Comment:
We can use a separate ticket to centralize these helper functions into ` org.apache.kafka.test.TestUtils`, would you mind filing a JIRA? Example functions like `getProducer/Consumer/StreamsConfig`, etc.